lib/src/lifecycle_module.dart

```diff
@@ -725,25 +735,36 @@ abstract class LifecycleModule extends SimpleModule
 
       ShouldUnloadResult shouldUnloadResult = shouldUnload();
       if (!shouldUnloadResult.shouldUnload) {
+        _state = _previousState;
+        _previousState = null;
+        _transition = null;
         // reject with shouldUnload messages
         throw new ModuleUnloadCanceledException(
             shouldUnloadResult.messagesAsString());
       }
-
       _willUnloadController.add(this);
-      await Future.wait(_childModules.map((child) => child.unload()));
+      List<Future<Null>> childUnloadFutures = <Future<Null>>[];
+      for (var child in _childModules.toList()) {
+        childUnloadFutures.add(child.unload());
+      }
+      _childModules.clear();
+      await Future.wait(childUnloadFutures);
       await onUnload();
       await _disposableProxy.dispose();
       if (_state == LifecycleState.unloading) {
         _state = LifecycleState.unloaded;
+        _previousState = null;
         _transition = null;
       }
       _didUnloadController.add(this);
-    } catch (error, stackTrace) {
+      await _didUnloadController.close();
+    } on ModuleUnloadCanceledException catch (error, stackTrace) {
       _didUnloadController.addError(error, stackTrace);
```

Comment:
Update: I've decided to change this behavior and consider it a bug. The key piece fact is that when a `shouldUnload` rejects an unload, the `willUnload` is not dispatched, meaning that if we add the error to the `didUnload` we're actually breaking the lifecycle itself. When the `shouldUnload` rejects, no streams will dispatch, but the error will be allowed to propagate to the `unload` call.